File: src/Aggregit.Web/Controllers/HomeController.fs
namespace Aggregit.Web.Controllers

open System
open System.Collections.Generic
open System.Linq
open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Aggregit.Web.Options
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Options
open System.Net.Cache
open System.Net.Http
open System.Text
open Newtonsoft.Json
open Aggregit.Web.ViewModels.Home

type GitHubPostData = { client_id:string; client_secret:string; code:string }
type IndexRouteValue = { accessToken:string }

type HomeController (serviceProvider:IServiceProvider) =
    inherit Controller()

    let gitHubApiOptions = serviceProvider.GetService<IOptions<GitHubApiOptions>>().Value

    member this.Index (accessToken:string) =
        let vm = new IndexViewModel ( AccessToken = accessToken )
        this.View(vm)
    
    member this.Error () =
        this.View()
    
    member this.Authorize (code:string) =
        let postAsync () = 
            async {
                let postData = { client_id = gitHubApiOptions.ClientId; client_secret = gitHubApiOptions.ClientSecret; code = code}
                let content = new StringContent(JsonConvert.SerializeObject(postData), Encoding.UTF8, "application/json")

                let httpClient = new System.Net.Http.HttpClient()
                let! response = httpClient.PostAsync("https://github.com/login/oauth/access_token", content) |> Async.AwaitTask
                response.EnsureSuccessStatusCode () |> ignore
                let! content = response.Content.ReadAsStringAsync() |> Async.AwaitTask
                return content
            }
        //TODO: Look for bad responses
        let response = postAsync() |> Async.RunSynchronously
        let strings = response.Split("&")
        let accessToken = strings.[0].Split("=").[1]

        this.RedirectToAction("Index", "Home", { accessToken = accessToken })

File: src/Aggregit.Web/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Aggregit.Web.Options;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Options;
using Microsoft.Extensions.DependencyInjection;
using Aggregit.Web.ViewModels.Home;
using System.Net.Http;
using Newtonsoft.Json;
using System.Text;

namespace Aggregit.Web.Controllers
{
    public class HomeController : Controller
    {
        private readonly GitHubApiOptions gitHubApiOptions;

        public HomeController(IServiceProvider serviceProvider)
        {
            gitHubApiOptions = serviceProvider.GetService<IOptions<GitHubApiOptions>>().Value;
        }

        [HttpGet]
        public IActionResult Index(string accessToken)
        {
            var vm = new IndexViewModel
            {
                AccessToken = accessToken
            };

            return View(vm);
        }

        public async Task<IActionResult> Authorize(string code)
        {
            var postData = new
            {
                client_id = gitHubApiOptions.ClientId,
                client_secret = gitHubApiOptions.ClientSecret,
                code = code
            };

            var content = new StringContent(JsonConvert.SerializeObject(postData), Encoding.UTF8, "application/json");
            var httpClient = new HttpClient();
            var response = await httpClient.PostAsync("https://github.com/login/oauth/access_token", content);
            response.EnsureSuccessStatusCode();

            var responseContent = await response.Content.ReadAsStringAsync();
            var strings = responseContent.Split("&");
            var accessToken = strings[0].Split("=")[1];

            //TODO: Check for failures
            //https://developer.github.com/apps/managing-oauth-apps/troubleshooting-oauth-app-access-token-request-errors/
            //if (string.Equals(accessToken, "bad_verification_code"))
            //{

            //}

            return RedirectToAction("Index", "Home", new { accessToken = accessToken });
        }

        public IActionResult Error()
        {
            return View();
        }
    }
}
